### PR TITLE
ArrayList solution

### PR DESCRIPTION
**Pull request by lskErixon**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,123 @@
...
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+
+    private static final int DEFAULT_INDEX = 0;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        this.elements = (T[]) new Object[DEFAULT_CAPACITY];
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:14]: Read [where ](https://stackoverflow.com/a/2411283) we should use `this`

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,123 @@
...
     public void set(T value, int index) {
+        checkIndex(index);
+        elements[index] = value;
+    }
 
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index);
+        }
     }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:59]: put private methods/members after public ones

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,123 @@
...
 
     @Override
     public void set(T value, int index) {
+        checkIndex(index);
+        elements[index] = value;
+    }
 
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index);
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:57]: include size too

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,123 @@
...
+            int newCapacity = (int) (elements.length * GROWTH_FACTOR);
+            T[] newArray = (T[]) new Object[newCapacity];
+            System.arraycopy(elements, DEFAULT_INDEX, newArray, DEFAULT_INDEX, size);
+            elements = newArray;
+        }
+    }
+
+    private int indexOf(T element) {
+        for (int i = 0; i < size; i++) {
+            if (elements[i] == null) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:111]: remove if-else
look up equals condition of Objects.equals and write this logic manually

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,123 @@
...
+        T removedElement = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (element == null) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:74]: remove if-else, just use correct equals condition that can handle nulls

**Review by okuzan — CHANGES_REQUESTED** [review]
Well done! Fix null handling

File: src/main/java/core/basesyntax/ArrayList.java (hunk 1 above)
Comment (lskErixon) [reply src/main/java/core/basesyntax/ArrayList.java:14]: thank you, it’s really useful for me

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,123 @@
...
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index);
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index);
+        }
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:65]: image

File: src/main/java/core/basesyntax/ArrayList.java (hunk 5 above)
Comment (Rommelua) [reply src/main/java/core/basesyntax/ArrayList.java:74]: we do not need this IF, delete it please

File: src/main/java/core/basesyntax/ArrayList.java (hunk 4 above)
Comment (Rommelua) [reply src/main/java/core/basesyntax/ArrayList.java:111]: not fixed

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Pull request closed without merging**